src/simulator/Simulator.tsx

```diff
@@ -0,0 +1,261 @@
+  useEffect(() => {
+    if (ref.current) {
+      const messageListener = (e: MessageEvent) => {
+        const simulator = ref.current!.contentWindow;
+        if (e.source === simulator) {
+          switch (e.data.kind) {
+            case "ready": {
+              setSensors(
+                Object.fromEntries(
+                  e.data.sensors.map((json: Sensor) => {
+                    return [json.id, json];
+                  })
+                )
+              );
+              while (readyCallbacks.current.length) {
+                readyCallbacks.current.pop()!();
+              }
+              break;
+            }
+            case "serial_output": {
+              // TODO: serial
+              break;
+            }
+          }
+        }
+      };
+      window.addEventListener("message", messageListener);
+      return () => {
+        window.removeEventListener("message", messageListener);
+      };
+    }
+  }, [ref, readyCallbacks]);
+
+  const onSensorChange = useCallback(
+    (id: string, value: number) => {
+      setSensors((sensors) => ({
+        ...sensors,
+        [id]: { ...sensors[id], value },
+      }));
+      const simulator = ref.current!.contentWindow!;
+      simulator.postMessage(
+        {
+          kind: "sensor_set",
+          sensor: id,
+          value,
+        },
+        "*"
+      );
+    },
+    [ref, setSensors]
+  );
+
+  const play = useCallback(async () => {
+    const filesystem: Record<string, Uint8Array> = Object.fromEntries(
+      await Promise.all(
+        fs.project.files.map(async (f) => [
+          f.name,
+          (await fs.read(f.name)).data,
+        ])
+      )
+    );
+    const simulator = ref.current!.contentWindow!;
+    simulator.postMessage(
+      {
+        kind: "flash",
+        filesystem,
+      },
+      "*"
+    );
+  }, [ref, fs]);
+
+  const stop = useCallback(async () => {
+    const simulator = ref.current!.contentWindow!;
+    simulator.postMessage(
+      {
+        kind: "serial_input",
+        // Ctrl-C to interrupt.
+        // A specific message would be useful as probably best to clear display etc. here.
+        data: `\x03`,
+      },
+      "*"
+    );
+  }, [ref]);
+
+  return {
+    play,
+    stop,
+    sensors: Object.values(sensors),
+    onSensorChange,
+  };
+};
+
+const Simulator = () => {
+  const ref = useRef<HTMLIFrameElement>(null);
+  const { play, stop, sensors, onSensorChange } = useSimulator(ref);
+  return (
+    <VStack spacing={5}>
+      <Box width="100%">
+        <AspectRatio ratio={400 / 321} width="100%">
+          <Box
+            ref={ref}
+            as="iframe"
+            // Very much a temporary / unsupported deployment!
+            src="https://stage-python-simulator.microbit.org/simulator.html"
+            title="Simulator"
+            frameBorder="no"
+            scrolling="no"
+          />
+        </AspectRatio>
+        {/* Margin hack until we remove space from iframe */}
```

Comment:
May we well fix this before merge.

Reply:
I I'll move the hack inside the iframe. The SVG has a hard to justify amount of whitespace in it but I want to avoid more changes to it for the moment.

Reply:
Ended up changing the SVG on https://github.com/microbit-foundation/micropython-microbit-v2-simulator/pull/7